bpo-36520: Email header folded incorrectly

websurfer5 opened the pull request:
Folding long email headers that contain multiple UTF-8 words in the first line results in corruption when UTF-8 words are found in subsequent lines. The fix is to reset the variable tracking the encoded word location in the line when a new line is started.
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-36520](https://bugs.python.org/issue36520) -->
https://bugs.python.org/issue36520
<!-- /issue-number -->

websurfer5 committed:
bpo-36520: reset the encoded word offset when starting a new
line during an email header folding operation

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

websurfer5 committed:
Merge branch 'master' of github.com:python/cpython into fix-issue-36520

websurfer5 committed:
Merge branch 'fix-issue-36520' of github.com:websurfer5/cpython into fix-issue-36520

maxking commented on Lib/test/test_email/test_message.py:
```diff
@@ -784,6 +784,80 @@ def test_str_defaults_to_utf8(self):
         self.assertEqual(str(m), 'Subject: unicöde\n\n')
 
+    def test_folding_with_utf8_encoding_1(self):
+        # issue #36520
```
It would be really helpful if these tests would add a one line comment about what are they actually testing. The name of the test uses numbers, so that doesn't help.

Something like,
`# Test header is folded correctly when maxlen falls in the middle of an encoded word.

websurfer5 replied:
I added descriptive comments to the test cases. IMO, meaningfully descriptive test names would be too lengthy for these tests.

maxking commented on Lib/email/_header_value_parser.py:
```diff
@@ -2661,6 +2661,7 @@ def _refold_parse_tree(parse_tree, *, policy):
             newline = _steal_trailing_WSP_if_exists(lines)
             if newline or part.startswith_fws():
                 lines.append(newline + tstr)
+                last_ew = None
```
Github won't let me comment on line not in this diff,  but Line 2683 of this diff does the similar wrapping with a newline and I was wondering if there could be a code path to trigger that resulting in a similar bug?

websurfer5 replied:
That code path never gets triggered when there are UTF-8 characters in the input. Line 2607 collapses the UnstructuredTokenList into a byte string containing the entire input text, and then lines 2609-2619 determine that the text contains UTF-8 characters and sets want_encoding to True. This always sends it down the code path that calls _fold_as_ew() to fold a line with encoded words and moves on to the next token regardless of whether individual tokens are ASCII or UTF-8.

maxking replied:
Ah, thanks for detailed explanation! My bad for not looking closely enough.

maxking commented on Misc/NEWS.d/next/Library/2019-05-28-02-37-00.bpo-36520.W4tday.rst:
```diff
@@ -0,0 +1 @@
+Lengthy email headers with UTF-8 characters are now properly encoded when they are folded.
```
Can you also add `Path by <your name>` to the end for correct attribution in changelog?

websurfer5 replied:
I added my name to the NEWS blurb.

maxking reviewed (commented):
Thanks for fixing this.

websurfer5 committed:
Merge branch 'master' of github.com:python/cpython into fix-issue-36520

websurfer5 committed:
Merge branch 'master' of github.com:python/cpython into fix-issue-36520

websurfer5 committed:
bpo-36520: add an additional test case, and provide descriptive
comments for the test_folding_with_utf8_encoding_* tests

websurfer5 committed:
bpo-36520: fix whitespace issue

warsaw commented on Lib/test/test_email/test_message.py:
```diff
@@ -784,6 +784,137 @@ def test_str_defaults_to_utf8(self):
         self.assertEqual(str(m), 'Subject: unicöde\n\n')
 
+    def test_folding_with_utf8_encoding_1(self):
+        # issue #36520
```
Shouldn't this say "bpo-36520"?  What would "issue #36520" refer to if we move to GitHub issues?

websurfer5 replied:
I have made the requested changes; please review again

warsaw commented on Lib/test/test_email/test_message.py:
```diff
@@ -784,6 +784,137 @@ def test_str_defaults_to_utf8(self):
+        m = EmailMessage()
+        m['Subject'] = 'Hello Wörld! Hello Wörld! '\
+                       'Hello Wörld! Hello Wörld!Hello Wörld!'
+        self.assertEqual(bytes(m), \
```
None of these backslashes are needed, since you're inside a parenthesized expression.

warsaw commented on Lib/test/test_email/test_message.py:
```diff
@@ -784,6 +784,137 @@ def test_str_defaults_to_utf8(self):
+        # word.
+
+        m = EmailMessage()
+        m['Subject'] = 'Hello Wörld! Hello Wörld! '\
```
Please put a space before the backslash.

warsaw commented on Lib/test/test_email/test_message.py:
```diff
@@ -784,6 +784,137 @@ def test_str_defaults_to_utf8(self):
+
+
+    def test_folding_with_utf8_encoding_2(self):
+        # issue #36520
```
bpo-36520

warsaw commented on Lib/test/test_email/test_message.py:
```diff
@@ -784,6 +784,137 @@ def test_str_defaults_to_utf8(self):
+        m = EmailMessage()
+        m['Subject'] = 'Hello Wörld! Hello Wörld! '\
+                       'Hello Wörlds123! Hello Wörld!Hello Wörld!'
+        self.assertEqual(bytes(m), \
```
Similar comment about all these (and subsequent) backslashes.

websurfer5 committed:
bpo-36520: changes per reviewer request -- remove extraneous
backslashes; add whitespace between terminating quotes and
line-continuation backslashes; use "bpo-" instead of
"issue #" in comments
